### PR TITLE
using multiple SchemaSwap annotations instead of the schell script

### LINKED ISSUE #10736: Use SchemaSwap instead of shell script for Realm CRD generatio
* Remove the [current workaround](https://github.com/keycloak/keycloak/blob/ad865e75c14f048e2de6748bab6af6c3b495de33/operator/patch-sources.sh) and rely purely on Fabric8 generator SchemaSwap.

### PR DESCRIPTION
**Pull request by shawkins**
I diffed the target/kubernetes/ crds before and after this change and there were no differences.  

However the swap logic complained about ComponentRepresentation - it may not be directly transitively reachable under KeycloakRealmImport.  I'm not sure if that's expected behavior.  cc @andreaTP 

Closes #10736

### CODE REVIEW

**Review by vmuzikar — COMMENTED** [review]
We should be also able to remove this:
https://github.com/keycloak/keycloak/blob/f61cf6660ac8b0a394b3303a034bdae1390d0b24/operator/pom.xml#L195-L210

**Comment by shawkins** [comment]
> We should be also able to remove this:

Ah right you were creating a local copy.  This has been removed now as well.

**Review by vmuzikar — DISMISSED** [review]

**Comment by vmuzikar** [comment]
LGTM as is (resulting CRD is identical to what we had before). Feel free to merge after the `ComponentRepresentation` issues were discussed resolved.

**Comment by shawkins** [comment]
To add more context about what I'm seeing, with 
```
@SchemaSwap(originalType = ComponentRepresentation.class, fieldName = "config", targetType = org.keycloak.representations.overrides.MultivaluedStringStringHashMap.class)
```
The build fails with
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:compile (default-compile) on project keycloak-operator: Fatal error compiling: java.lang.IllegalArgumentException: Unmatched SchemaSwaps: @SchemaSwap(originalType=org.keycloak.representations.idm.ComponentRepresentation, fieldName="config", targetType=org.keycloak.representations.overrides.MultivaluedStringStringHashMap) on org.keycloak.operator.crds.v2alpha1.realmimport.KeycloakRealmImport
```

> Feel free to merge after the ComponentRepresentation issues were discussed.

Let's hold offf on that.  With the commit as is, with the annotation referencing ComponentRepresentation removed, I'm now consistently seeing:

```
                           decisionStrategy:
                              enum:
                              - stableIndex
                              - CONSENSUS
                              - UNANIMOUS
                              - AFFIRMATIVE
                              - ENUM$VALUES
                              type: string
                            name:
                              type: string
                            policyEnforcementMode:
                              enum:
                              - stableIndex
                              - PERMISSIVE
                              - ENFORCING
                              - ENUM$VALUES
                              - DISABLED
                              type: string
```

Where **ENUM$VALUES** is new as a possible enum value in the realmimports crd on all enums.  At first glance that does seem like a crd generator bug.  None of these enums are referenced in any way by ComponentRepresentation, so it doesn't seem directly related to that schemaswap not being used.  cc @andreaTP 

In fact it seems to make no difference to the crd whether any of the config SchemaSwaps are used - with 
```
@SchemaSwap(originalType = CredentialRepresentation.class, fieldName = "config", targetType = org.keycloak.representations.overrides.MultivaluedStringStringHashMap.class)
@SchemaSwap(originalType = ComponentExportRepresentation.class, fieldName = "config", targetType = org.keycloak.representations.overrides.MultivaluedStringStringHashMap.class)
```
removed you still get the same crd's generated - is it possible that the crd logic now appropriately handles the typing and no longer requires MultivaluedStringStringHashMap?

**Comment by andreaTP** [comment]
Multiple questions in this thread, let me know if I miss something!

> *ENUM$VALUES*

This seems like a bug, can you reduce it and report it in the client repo?

> `ComponentRepresentation` - it may not be directly transitively reachable under KeycloakRealmImport.

This is totally possible, the weak error-handling capabilities of the bash script might have hidden a drift here.

> is it possible that the crd logic now appropriately handles the typing and no longer requires MultivaluedStringStringHashMap?

I strongly doubt this is possible, what's the `yaml` shape of the generated `config` field with vs. without the annotation?
From a first quick check seems like `ComponentExportRepresentation` is not reachable as well, the fact that `SchemaSwap` is not complaining might be a bug in the erroneous condition detection logic (we have reworked it various times already ...).

**Comment by shawkins** [comment]
> This seems like a bug, can you reduce it and report it in the client repo?

Captured as https://github.com/fabric8io/kubernetes-client/issues/5228 - not sure how to induce it yet or why it only is happening for me with SchemaSwap.

> I strongly doubt this is possible, what's the yaml shape of the generated config field with vs. without the annotation?
From a first quick check seems like ComponentExportRepresentation is not reachable as well, the fact that SchemaSwap is not complaining might be a bug in the erroneous condition detection logic (we have reworked it various times already ...).

Backing the SchemaSwap change out and modifying patch-sources to remove all of the config MultivaluedStringStringHashMap patches shows the same behavior - the generated crd is identical either way.  

Every config: I see in the realmimport crd is simply an object:
```
config:
  additionalProperties:
    type: string
  type: object
```
I'll try to track down how these map back to the classes that are referenced in patch-sources / SchemaSwap

**Comment by shawkins** [comment]
If there isn't already something, we may need to consider having the blessed version of the crds checked in somewhere so that any changes can be validated.  I realize this will add an extra step anytime there's a modification that defeats some of the point of it being generated, but it seems like we may need guard against the unexpected more.

**Comment by andreaTP** [comment]
This used to be the place for them: https://github.com/keycloak/keycloak-k8s-resources/tree/main/kubernetes

**Comment by shawkins** [comment]
> Quick follow-up regarding ENUM$VALUES, despite it being a bug, if it's not a regression, shouldn't block this PR.

I agree - but there should be some kind of post check to verify we're not accidentally introducing those into the CRD. 

> This used to be the place for them: https://github.com/keycloak/keycloak-k8s-resources/tree/main/kubernetes

Since this doesn't appear to be kept in sync should there just be a maven exec or something to check for ENUM$VALUES?

Do you have any thoughts on the config schemas - or should that just be tracked with a fresh issue since it existed before this pr?

**Comment by abstractj** [comment]
> Do you have any thoughts on the config schemas - or should that just be tracked with a fresh issue since it existed before this pr?

@shawkins from my perspective, it is totally fine to have a follow-up issue, considering that the issue mentioned was not introduced by the changes proposed here.

**Comment by shawkins** [comment]
> @shawkins from my perspective, it is totally fine to have a follow-up issue, considering that the issue mentioned was not introduced by the changes proposed here.

@abstractj Ok, I made one final update to this pr to use an ant task to remove the potentially invalid enum entries.  If that seems unnecessary I can back that out.  The follow ups are:
- #20935
- #20936
- #20937

File: operator/pom.xml [hunk 1]
```diff
@@ -288,6 +254,27 @@
                     </systemPropertyVariables>
                 </configuration>
             </plugin>
+            
+            <plugin>
+                <artifactId>maven-antrun-plugin</artifactId>
```
Comment (ASzc) [inline operator/pom.xml:259]: Fully specified GAVs are always preferred (minimum of `groupId` and `artifactId`). Although I can find lots of other places in the code already, where we're using antrun without a `groupId` set, so I guess it's not breaking anything? You will inherit jboss-parent's specified version of 1.8 by default, so no need to add a `version` tag in this case.

Comment (shawkins) [reply operator/pom.xml:259]: > You will inherit jboss-parent's specified version of 1.8 by default, so no need to add a version tag in this case.

Yes that's what I saw in the ide.  Ideally this task will go away shortly via #20935

**Review by Pepo48 — APPROVED** [review]
LGTM as well. I agree that merging this shouldn't be blocked by the follow-up bug on the client side.

@shawkins nice job. The shell script generation is finally gone. Thank you :)

**Review by abstractj — APPROVED** [review]
Thank you @shawkins

**Pull request merged**